### PR TITLE
Support type coercion for join on columns

### LINKED ISSUE #2877: hash_join panics when join keys have different data types
**Describe the bug**

There are several places where hash_join can panic in version 9.0.0 when I run SQL queries.

```
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', /home/andy/.cargo/registry/src/github.com-1ecc6299db9ec823/datafusion-9.0.0/src/physical_plan/hash_join.rs:973:17
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', /home/andy/.cargo/registry/src/github.com-1ecc6299db9ec823/datafusion-9.0.0/src/physical_plan/hash_join.rs:979:17
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', /home/andy/.cargo/registry/src/github.com-1ecc6299db9ec823/datafusion-9.0.0/src/physical_plan/hash_join.rs:1045:17
```

**To Reproduce**
I am experimenting with SQL query fuzzing. Here is one query that failed:

```sql
SELECT _c438, _c439, _c440, _c441, _c442, _c443
FROM (
  (
    (SELECT test1.c0 AS _c438, test1.c1 AS _c439, test1.c2 AS _c440, test1.c3 AS _c441, test1.c4 AS _c442, test1.c5 AS _c443
      FROM (test1))
    FULL JOIN
    (SELECT test1.c0 AS _c444, test1.c1 AS _c445, test1.c2 AS _c446, test1.c3 AS _c447, test1.c4 AS _c448, test1.c5 AS _c449
      FROM (test1))
    ON _c438 = _c446)
  RIGHT JOIN
  (SELECT test0.c0 AS _c450, test0.c1 AS _c451, test0.c2 AS _c452, test0.c3 AS _c453, test0.c4 AS _c454, test0.c5 AS _c455
    FROM (test0))
  ON _c441 = _c451);
```

```
ArrowError(ExternalError("Arrow error: External error: Execution error: Join Error: task 5672 panicked"))
```

Data files are here - https://github.com/andygrove/sqlfuzz/tree/main/testdata

**Expected behavior**
Should not panic

**Additional context**

### PR DESCRIPTION
**Pull request by ygf11**
# Which issue does this PR close?

Closes #2877.

# Rationale for this change
Datafusion support type coercion for expressions, it should also work for join keys.

# What changes are included in this PR?
In type coercion rule, if the plan is a join:

* Add projection for the join input if its column need type coercion.
* Since one column may be casted one more time, the casted keys will have an alias. 
 
# Are these changes tested?
Yes, unit test will test it.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/type_coercion.rs [hunk 1]
```diff
@@ -1114,4 +1293,145 @@ mod test {
...
+            None,
+        )?;
+
+        let expected = "Projection: t1.a, t1.b, t1.c, t2.a, t2.b, t2.c\
+        \n  Inner Join: t1.a * UInt8(2)#0 = t2.b, t1.a * UInt8(2)#1 = t2.c\
+        \n    Projection: t1.a, t1.b, t1.c, t1.a * UInt8(2), CAST(t1.a * UInt8(2) AS Int32) AS t1.a * UInt8(2)#0, CAST(t1.a * UInt8(2) AS Int64) AS t1.a * UInt8(2)#1\
+        \n      Projection: t1.a, t1.b, t1.c, t1.a * CAST(UInt8(2) AS UInt16)\
+        \n        TableScan: t1\
+        \n    TableScan: t2";
+        let rule = TypeCoercion::new();
```
Comment (ygf11) [inline datafusion/optimizer/src/type_coercion.rs:1430]: In this test case, the column happens twice and both need type coercion, I give them an alias to distinguish from each other.

**Comment by ygf11** [comment]
@alamb @mingmwang @liukun4515 PTAL.

**Comment by liukun4515** [comment]
> @alamb @mingmwang @liukun4515 PTAL.

Thanks @ygf11 I will try to find time to review it for the next few days.

**Comment by ygf11** [comment]
Marked this pr as draft, wait #4389 fixed.

**Comment by ygf11** [comment]
@liukun4515 @andygrove I create a new one #4666.

**Pull request closed without merging**